File: framework/languages/cpp/e2e/ToActorMessaging/Server/Actor/main.cpp
/* SPDX-License-Identifier: FSL-1.1-ALv2 */

#include "../../Shared/messages.hpp"
#include "../../Shared/configuration.hpp"

#include <zlink/framework.hpp>
#include <zlink/locations/redis.hpp>

#include <chrono>
#include <mutex>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace e2e = zlink::e2e::to_actor_messaging;

namespace
{

void add_redis_location_store (zlink::framework::zlink_framework_options_t &framework,
                               const e2e::redis_configuration_t &redis)
{
    framework.add_location_store (
      std::make_shared<zlink::framework::redis::redis_location_store_t> (
        zlink::framework::redis::redis_location_options_t{
          .connection_string = redis.endpoint, .key_prefix = redis.key_prefix}));
    auto &locations = framework.configure_locations ();
    locations.heartbeat_interval = std::chrono::seconds (1);
    locations.owner_lease_ttl = std::chrono::seconds (3);
    locations.polling_interval = std::chrono::milliseconds (500);
}

class evidence_store_t
{
  public:
    void append (e2e::actor_evidence_t entry)
    {
        std::lock_guard<std::mutex> lock (_mutex);
        _entries.push_back (std::move (entry));
    }

    std::vector<e2e::actor_evidence_t> all () const
    {
        std::lock_guard<std::mutex> lock (_mutex);
        return _entries;
    }

  private:
    mutable std::mutex _mutex;
    std::vector<e2e::actor_evidence_t> _entries;
};

class to_actor_e2e_actor_t : public zlink::framework::actor_t
{
  public:
    explicit to_actor_e2e_actor_t (
      zlink::framework::actor_context_t context) :
        _actor_id (context.actor_ref ().actor_id ().value ()),
        _context (std::move (context))
    {
    }
    zlink::framework::actor_context_t &context () noexcept override
    { return _context; }
    const zlink::framework::actor_context_t &context () const noexcept override
    { return _context; }

    const std::string &actor_id () const { return _actor_id; }

  private:
    std::string _actor_id;
    zlink::framework::actor_context_t _context;
};

class to_actor_e2e_actor_factory_t final
    : public zlink::framework::actor_factory_t<
        to_actor_e2e_actor_t>
{
  public:
    zlink::framework::task_t<std::shared_ptr<
      to_actor_e2e_actor_t>>
    create (zlink::framework::actor_context_t context,
            std::stop_token) override
    {
        co_return std::make_shared<to_actor_e2e_actor_t> (
          std::move (context));
    }
};

class to_actor_e2e_spot_t
    : public zlink::framework::entry_spot_t<to_actor_e2e_actor_t>
{
  public:
    to_actor_e2e_spot_t (zlink::framework::entry_spot_context_t context,
                         evidence_store_t &evidence) :
        _evidence (evidence), _context (std::move (context))
    {
    }

    zlink::framework::entry_spot_context_t &context () noexcept override
    {
        return _context;
    }

    const zlink::framework::entry_spot_context_t &context () const noexcept override
    {
        return _context;
    }

    void configure () override
    {
        _context.handlers ().add_actor_send<&to_actor_e2e_spot_t::on_notify> (
          e2e::actor_notify_t::packet_name);
        _context.handlers ().add_actor_request<&to_actor_e2e_spot_t::on_ask> (
          e2e::actor_ask_t::packet_name);
    }

    zlink::framework::task_t<zlink::framework::actor_create_response_t>
    on_create_actor (to_actor_e2e_actor_t &actor,
                     const zlink::framework::message_t &) override
    {
        _evidence.append ({"create", actor.actor_id (), "create", "created"});
        co_return zlink::framework::actor_create_response_t::accept ();
    }

    zlink::framework::task_t<zlink::framework::spot_actor_join_result_t>
    on_actor_join (std::string_view,
                   const zlink::framework::message_t &) override
    {
        co_return zlink::framework::spot_actor_join_result_t::accept ();
    }

    zlink::framework::task_t<void>
    on_actor_joined (to_actor_e2e_actor_t &actor) override
    {
        _evidence.append ({"join", actor.actor_id (), "join", "joined"});
        co_return;
    }

    zlink::framework::task_t<void>
    on_leave_actor (to_actor_e2e_actor_t &) override
    {
        co_return;
    }

    void on_notify (to_actor_e2e_actor_t &actor,
                    zlink::framework::spot_actor_send_context_t &,
                    const e2e::actor_notify_t &message)
    {
        _evidence.append ({message.scenario, actor.actor_id (), "send", message.value});
    }

    zlink::framework::task_t<e2e::actor_reply_t>
    on_ask (to_actor_e2e_actor_t &actor,
            zlink::framework::spot_actor_request_context_t &,
            const e2e::actor_ask_t &message)
    {
        _evidence.append ({message.scenario, actor.actor_id (), "request", message.value});
        auto reply = e2e::actor_reply_t{message.scenario, actor.actor_id (),
                                        "reply:" + message.value};
        if (message.scenario == "TA-B1-destroy" || message.scenario == "TA-A4-destroy"
            || message.scenario == "TA-B2-destroy") {
            co_await _context.destroy_actor (actor);
        }
        co_return reply;
    }

  private:
    evidence_store_t &_evidence;
    zlink::framework::entry_spot_context_t _context;
};

class ensure_actor_handler_t
{
  public:
    using dependency_types =
      zlink::framework::dependency_list_t<zlink::framework::session_actor_manager_t,
                                         e2e::actor_configuration_t>;
    using request_type = e2e::actor_call_request_t;
    using reply_type = e2e::actor_call_response_t;

    ensure_actor_handler_t (zlink::framework::session_actor_manager_t actors,
                            e2e::actor_configuration_t &configuration) :
        _actors (std::move (actors)), _configuration (configuration)
    {
    }

    e2e::actor_call_response_t handle (const e2e::actor_call_request_t &request)
    {
        try {
            auto actor = _actors.get_or_create (e2e::actor_type_name, request.actor_id,
                                                std::string ("create"));
            if (!actor) {
                throw *actor.error ();
            }
            auto bound = _actors.bind_or_get (actor.value ().ref ()).submit ().result ();
            if (!bound) {
                throw *bound.error ();
            }
            auto joined = bound.value ()
                            .context ()
                            .join_entry_spot (
                              zlink::framework::node_rid_t::from_string (
                                _configuration.node_rid),
                              zlink::framework::message_t {})
                            .async ()
                            .result ();
            if (!joined) {
                throw *joined.error ();
            }
            return {request.scenario, request.actor_id, "ensured", ""};
        }
        catch (const zlink::framework::framework_exception_t &error) {
            return {request.scenario, request.actor_id, "",
                    std::to_string (static_cast<int> (error.kind ())) + ":" + error.what ()};
        }
    }

  private:
    zlink::framework::session_actor_manager_t _actors;
    e2e::actor_configuration_t &_configuration;
};

class evidence_handler_t
{
  public:
    using dependency_types = zlink::framework::dependency_list_t<evidence_store_t>;

    explicit evidence_handler_t (evidence_store_t &evidence) : _evidence (evidence) {}

    zlink::framework::http_response_t handle (const zlink::framework::http_request_t &)
    {
        zlink::framework::http_response_t response;
        response.body = nlohmann::json (_evidence.all ()).dump ();
        return response;
    }

  private:
    evidence_store_t &_evidence;
};

class push_actor_handler_t
{
  public:
    using dependency_types =
      zlink::framework::dependency_list_t<zlink::framework::session_actor_manager_t>;
    using request_type = e2e::actor_call_request_t;
    using reply_type = e2e::actor_call_response_t;

    explicit push_actor_handler_t (zlink::framework::session_actor_manager_t &actors) :
        _actors (actors)
    {
    }

    e2e::actor_call_response_t handle (const e2e::actor_call_request_t &request)
    {
        const auto actor = _actors.find (request.actor_id);
        if (!actor) {
            return {request.scenario, request.actor_id, "", "actor_route_not_found"};
        }
        actor->bound_session ()
          .send (e2e::actor_push_notify_t{request.scenario, request.actor_id, request.value})
          .submit ();
        return {request.scenario, request.actor_id, "pushed", ""};
    }

  private:
    zlink::framework::session_actor_manager_t &_actors;
};

} // namespace

int main (int argc, char **argv)
{
    auto app = zlink::framework::app_t::create ();
    const auto configuration =
      e2e::load_role_configuration<e2e::actor_configuration_t> (app, argc, argv);
    app.logging ().use_file (configuration.log_dir + "/actor.log").set_min_level (
      zlink::framework::log_level_t::debug);
    app.add_zlink_framework ([configuration] (
                              zlink::framework::zlink_framework_options_t &framework) {
        auto evidence = std::make_unique<evidence_store_t> ();
        auto *evidence_ptr = evidence.get ();
        framework.configure_dispatch ()
          .message_flow (zlink::framework::message_flow_log_mode_t::key_transitions)
          .trace_log_file (configuration.log_dir + "/actor-flow.log")
          .trace_label ("cpp-to-actor-actor");
        framework.services ().add_singleton<evidence_store_t> (std::move (evidence));
        framework.services ().add_singleton<e2e::actor_configuration_t> (
          std::make_unique<e2e::actor_configuration_t> (configuration));
        add_redis_location_store (framework, configuration.redis);
        auto mesh = framework.add_route_mesh (e2e::spot_mesh_name)
          .listen (configuration.spot_endpoint)
          .set_routing_id (zlink::routing_id_t::from (configuration.node_rid))
          .add_entry_spot<to_actor_e2e_spot_t> (
            [evidence_ptr] (zlink::framework::entry_spot_context_t context) {
                return std::make_shared<to_actor_e2e_spot_t> (
                  std::move (context), *evidence_ptr);
            })
          .add_actor_factory<
            to_actor_e2e_actor_t,
            to_actor_e2e_actor_factory_t> (
            e2e::actor_type_name,
            std::make_shared<to_actor_e2e_actor_factory_t> (),
            [] (auto &factory) { factory.disable_relocation (); });
        if (!configuration.caller_spot_endpoint.empty ()) {
            mesh.peer_connections ().connect (
              zlink::routing_id_t::from (configuration.caller_rid),
              configuration.caller_spot_endpoint);
        }
        framework.http ()
          .listen (configuration.http_endpoint)
          .map_health ("/health")
          .map_get<evidence_handler_t> ("/evidence")
          .map_post<push_actor_handler_t> ("/push")
          .map_post<ensure_actor_handler_t> ("/ensure");
    });
    return app.run (argc, argv);
}

File: framework/languages/node/e2e/ToActorMessaging/Server/Actor/main.ts
import fs from 'node:fs';
import path from 'node:path';
import { Inject, Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import {
  ZLinkMessageFlowLogMode,
  ZLinkMessage,
  ZLinkFrameworkErrorKind,
  ZLinkFrameworkException,
  type ZLinkActorCreateResponse,
  type ActorRef,
  type ZLinkActor,
  type ZLinkActorClient,
  type ZLinkActorContext,
  type ZLinkActorFactory,
  type ZLinkEntrySpot,
  type ZLinkEntrySpotContext,
  type ZLinkMessageContext,
  type ZLinkLocationRuntimeQuery,
  type ZLinkRouteMeshRuntime
} from '@zlink-systems/framework';
import {
  ZLINK_ACTOR_MANAGER,
  ZLINK_ACTOR_CLIENT,
  ZLINK_LOCATION_RUNTIME_QUERY,
  ZLINK_ROUTE_MESH_RUNTIME,
  ZLinkModule,
  zlinkEntrySpotActorRequestHandler,
  zlinkEntrySpotActorSendHandler,
  zlinkFramework
} from '@zlink-systems/nestjs';
import type { ZLinkActorManager } from '@zlink-systems/framework';
import { createRedisLocationStore, locationMessagingOptions } from '../../Shared/location-store';
import {
  PacketNames,
  type ActorAsk,
  type ActorNotify,
  ActorPushNotify,
  type ActorPushReq,
  type ActorRefPayload,
  type ActorReply
} from '../../Shared/messages';
import { EvidenceStore } from './evidence-store';
import { closeHttpServer, startHttpServer } from '../Support/http-server';
import { createLocationTopologyRoute } from '../Support/location-topology-route';
import { TO_ACTOR_OPTIONS, createToActorConfigurationModule } from '../../configuration';
import type { ServerOptions } from '../../configuration';

let options: ServerOptions;
let evidence: EvidenceStore;
let stopping = false;

class TestActor implements ZLinkActor {
  constructor(
    readonly actorId: string,
    readonly context: ZLinkActorContext
  ) {}
}

class TestActorFactory implements ZLinkActorFactory {
  async create(context: ZLinkActorContext): Promise<TestActor> {
    return new TestActor(context.actorId, context);
  }
}

class TestEntrySpot implements ZLinkEntrySpot<TestActor> {
  readonly context!: ZLinkEntrySpotContext<TestActor>;
  private readonly actors = new Map<string, TestActor>();
  private readonly pendingDestroys = new Map<string, Promise<void>>();

  constructor(
    @Inject(ZLINK_ACTOR_CLIENT) private readonly actorClient: ZLinkActorClient
  ) {}

  async onCreateActor(actor: TestActor, _request: ZLinkMessage): Promise<ZLinkActorCreateResponse> {
    this.actors.set(actor.actorId, actor);
    evidence.append({ scenario: 'create', actorId: actor.actorId, kind: 'create', value: 'created' });
    return { accepted: true };
  }

  async onJoinedActor(_actor: TestActor): Promise<void> {}

  async onLeaveActor(_actor: TestActor): Promise<void> {}

  async onDisconnectActor(_actor: TestActor): Promise<void> {}

  async destroy(actorId: string): Promise<void> {
    const actor = this.actors.get(actorId);
    if (actor === undefined) {
      throw new Error(`Actor '${actorId}' is not active.`);
    }
    await this.actorClient
      .requestToActor(actorId, new DestroySelfReq())
      .submit<{ readonly scheduled: boolean }>();
    await this.pendingDestroys.get(actorId);
    this.pendingDestroys.delete(actorId);
    this.actors.delete(actorId);
  }

  scheduleDestroy(actor: TestActor): void {
    const operation = this.context.runIoWorker(async () => true).submit().then(async () => {
      await this.context.destroyActor(actor);
    });
    this.pendingDestroys.set(actor.actorId, operation);
  }
}

class DestroySelfReq {}

@zlinkEntrySpotActorRequestHandler({
  actor: () => TestActor,
  entrySpot: () => TestEntrySpot,
  packetName: 'DestroySelfReq'
})
class DestroySelfHandler {
  async handle(
    spot: TestEntrySpot,
    actor: TestActor,
    _context: ZLinkMessageContext,
    _request: DestroySelfReq
  ): Promise<{ readonly scheduled: boolean }> {
    spot.scheduleDestroy(actor);
    return { scheduled: true };
  }
}

@zlinkEntrySpotActorSendHandler({
  actor: () => TestActor,
  entrySpot: () => TestEntrySpot,
  packetName: PacketNames.actorNotify
})
class NotifyHandler {
  async handle(_spot: TestEntrySpot, actor: TestActor, _context: ZLinkMessageContext, message: ActorNotify): Promise<void> {
    evidence.append({ scenario: message.scenario, actorId: actor.actorId, kind: 'send', value: message.value });
  }
}

@zlinkEntrySpotActorRequestHandler({
  actor: () => TestActor,
  entrySpot: () => TestEntrySpot,
  packetName: PacketNames.actorAsk
})
class AskHandler {
  async handle(_spot: TestEntrySpot, actor: TestActor, _context: ZLinkMessageContext, request: ActorAsk): Promise<ActorReply> {
    if (request.value === 'throw') {
      throw new Error('to-actor handler exception');
    }
    evidence.append({ scenario: request.scenario, actorId: actor.actorId, kind: 'request', value: request.value });
    return { scenario: request.scenario, actorId: actor.actorId, value: `reply:${request.value}` };
  }
}

@zlinkEntrySpotActorRequestHandler({
  actor: () => TestActor,
  entrySpot: () => TestEntrySpot,
  packetName: PacketNames.actorPush
})
class PushHandler {
  async handle(_spot: TestEntrySpot, actor: TestActor, _context: ZLinkMessageContext, request: ActorPushReq): Promise<ActorReply> {
    await actor.context.boundSession
      .send(new ActorPushNotify(request.scenario, actor.actorId, request.value))
      .submit();
    evidence.append({ scenario: request.scenario, actorId: actor.actorId, kind: 'push', value: request.value });
    return { scenario: request.scenario, actorId: actor.actorId, value: `pushed:${request.value}` };
  }
}

class ActorModule {}
const configuration = createToActorConfigurationModule();
Module({
  imports: [
    configuration,
    ZLinkModule.forRootFactory({
      imports: [configuration],
      inject: [TO_ACTOR_OPTIONS],
      useFactory: (value: unknown) => {
        options = value as ServerOptions;
        fs.mkdirSync(options.logDir, { recursive: true });
        evidence = new EvidenceStore(options.evidenceFile ?? path.join(options.logDir, 'actor.evidence.log'));
        const builder = zlinkFramework();
        builder
          .addLocationStore(createRedisLocationStore({
            redisEndpoint: options.redisEndpoint,
            redisKeyPrefix: options.redisKeyPrefix
          }));
        locationMessagingOptions(builder.configureLocations());
        builder
          .configureDispatch()
          .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
          .traceLogFile(path.join(options.logDir, 'actor-flow.log'))
          .traceLabel(options.rid);
        const mesh = builder
          .addRouteMesh('to-actor')
          .listen(options.routerEndpoint).routingId(options.rid);
        const objectServer = mesh.objects().server();
        objectServer.addEntrySpot(TestEntrySpot);
        objectServer.addActorFactory(
          'test-actor',
          TestActorFactory,
          (factory) => factory.disableRelocation()
        );
        mesh.channel('to-actor').server();
        return builder.build();
      }
    })
  ],
  providers: [TestActorFactory, TestEntrySpot, NotifyHandler, AskHandler, PushHandler, DestroySelfHandler]
})(ActorModule);

async function main(): Promise<void> {
  const app = await NestFactory.createApplicationContext(ActorModule, { logger: false, abortOnError: false });
  const actors = app.get(ZLINK_ACTOR_MANAGER, { strict: false }) as ZLinkActorManager;
  const entrySpot = app.get(TestEntrySpot, { strict: false });
  const locations = app.get(ZLINK_LOCATION_RUNTIME_QUERY, { strict: false }) as ZLinkLocationRuntimeQuery;
  const routeMeshRuntime = app.get(ZLINK_ROUTE_MESH_RUNTIME, { strict: false }) as ZLinkRouteMeshRuntime;
  const server = await startHttpServer(options.httpUrl, [
    { method: 'GET', path: '/health', handle: () => ({ status: 'ok' }) },
    createLocationTopologyRoute(locations, routeMeshRuntime),
    { method: 'GET', path: '/evidence', handle: () => evidence.all() },
    {
      method: 'POST',
      path: '/actors/ta-a1/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-a1');
        return { actorId: 'ta-a1', actor: actorSnapshot(actor) };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-a2/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-a2');
        return { actorId: 'ta-a2', actor: actorSnapshot(actor) };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-a3/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-a3');
        return { actorId: 'ta-a3', actor: actorSnapshot(actor) };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-a4/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-a4');
        return { actorId: 'ta-a4', actor: actorSnapshot(actor) };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-b1-reference/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-b1-reference');
        return { actorId: 'ta-b1-reference', actor: actorSnapshot(actor) };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-a4/destroy',
      handle: async () => {
        await entrySpot.destroy('ta-a4');
        return { actorId: 'ta-a4', status: 'destroyed' };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-b1-reference/destroy',
      handle: async () => {
        await entrySpot.destroy('ta-b1-reference');
        return { actorId: 'ta-b1-reference', status: 'destroyed' };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-b2/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-b2');
        return { actorId: 'ta-b2', actor: actorSnapshot(actor) };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-b2/destroy',
      handle: async () => {
        await entrySpot.destroy('ta-b2');
        return { actorId: 'ta-b2', status: 'destroyed' };
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-b2/destroy-ref',
      handle: async (body) => {
        try {
          const destroyed = await actors.destroy(actorRefFromSnapshot(body as ActorRefPayload));
          return { actorId: 'ta-b2', status: destroyed ? 'destroyed' : 'not-found' };
        } catch (error) {
          return {
            actorId: 'ta-b2',
            status: 'failed',
            errorKind: error instanceof ZLinkFrameworkException
              ? ZLinkFrameworkErrorKind[error.kind]
              : error instanceof Error
                ? error.name
                : String(error)
          };
        }
      }
    },
    {
      method: 'POST',
      path: '/actors/ta-b3/ensure',
      handle: async () => {
        const actor = await ensureActor(actors, 'ta-b3');
        return { actorId: 'ta-b3', actor: actorSnapshot(actor) };
      }
    },
    { method: 'POST', path: '/shutdown', handle: () => { stopping = true; return { status: 'stopping' }; } }
  ]);

  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}

function actorSnapshot(actor: ActorRef): ActorRefPayload {
  return {
    nodeRid: String(actor.nodeRid),
    actorId: actor.actorId,
    objectGeneration: actor.objectGeneration.toString(),
    meshName: actor.meshName
  };
}

function actorRefFromSnapshot(actor: ActorRefPayload): ActorRef {
  return {
    nodeRid: actor.nodeRid,
    actorId: actor.actorId,
    objectGeneration: BigInt(actor.objectGeneration),
    meshName: actor.meshName
  };
}

async function ensureActor(actors: ZLinkActorManager, actorId: string): Promise<ActorRef> {
  const result = await actors
    .getOrCreate(actorId, 'test-actor')
    .inMesh('to-actor')
    .submit();
  if (result.status === 'rejected') {
    throw new Error(`Actor '${actorId}' creation was rejected.`);
  }
  return result.actor;
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});
